### PR TITLE
Adds the ability to explicitly configure any Zipkin sender without flags

### PR DESCRIPTION
**Pull request by adriancole**
Not all configuration is available with Finagle flags. This adds a small
wrapper that allows someone to explicitly configure whatever they want.

See https://groups.google.com/forum/#!topic/zipkin-user/AxOo1Jyt2jI

NOTE: unit tests pending until it is verified people want this and would
use it. cc @crispywalrus

### CODE REVIEW

File: core/src/main/java/zipkin2/finagle/BasicZipkinTracer.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ *   .overrides(overrides)
+ *   .encoding(Encoding.PROTO3)
+ *   .build();
+ *
+ * // Now, use it here, but don't forget to close the sender!
+ * tracer = ZipkinTracer.create(sender);
+ * }</pre>
+ *
+ */
+public final class BasicZipkinTracer extends ZipkinTracer {
```
Comment (adriancole) [inline core/src/main/java/zipkin2/finagle/BasicZipkinTracer.java]: Flexible on the name of this type..

**Comment by adriancole** [comment]
If someone wants to try this branch you can setup maven like this

https://jitpack.io/#openzipkin/zipkin-finagle/basic-tracer-SNAPSHOT

or something like this:

```bash
TAG=basic-tracer-SNAPSHOT
curl -sSL https://jitpack.io/com/github/openzipkin/zipkin-finagle/zipkin-finagle_2.12/${TAG}/zipkin-finagle_2.12-${TAG}.jar > zipkin-finagle_2.12.jar
```

**Comment by crispywalrus** [comment]
Insofar as this simply adds the ability to create a new tracer with an externally constructed sender it seems fine. Because the passed in sender does get managed the same as an internally constructed sender I would suggest buffing the docs to make that very explicit.

File: core/src/main/java/zipkin2/finagle/BasicZipkinTracer.java (hunk 1 above)
Comment (crispywalrus) [reply core/src/main/java/zipkin2/finagle/BasicZipkinTracer.java]: `Basic` seems wrong to me. I think you meant it as "basically you have to do all of this so it's all less automatic" but I initially took it as a more basic (i.e. simpler) mechanism when it's the opposite. I like the builder and it makes me wonder if the idea can't be lifted up the stack more than one `newBuilder` method

File: core/src/main/java/zipkin2/finagle/BasicZipkinTracer.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  /** Note you must close the supplied sender externally, after this instance is closed. */
+  public static ZipkinTracer create(Sender sender) {
+    return new Builder(sender).build();
```
Comment (crispywalrus) [inline core/src/main/java/zipkin2/finagle/BasicZipkinTracer.java]: this is probably over pedantic but `return newBuilder(sender).build();` seems ever so slightly better

**Comment by adriancole** [comment]
basic is indeed a bad name. explicit I considered but that word has
multiple meanings :)

I considered putting this as a nested detail of the Tracer class.. the only
problem is that subtypes would inherit it. for example the static
newBuilder method would be visible on KafkaTracer eventhough it isnt
relevant. might be an ok tradeoff wdyt?

**Comment by crispywalrus** [comment]
So I like the idea of pushing this up into Tracer despite it not having universal currency in the the implementations. It's always a choice to use such a low level API and you wouldn't unless it has some direct benefit to you. Plus you get to push the naming problem away ... `Manual`?

**Comment by adriancole** [comment]
> So I like the idea of pushing this up into Tracer despite it not having
> universal currency in the the implementations. It's always a choice to use
> such a low level API and you wouldn't unless it has some direct benefit to
> you. Plus you get to push the naming problem away ... Manual?
>

Manual seems fine, but actually in practice I do like your idea better, to
just pull it into the Tracer. Another reason is that few instantiate these
objects anyway. The awkwardness of the design is a testament to this.. for
example these types have to satisfy service loader.. most people using
service loader also means most won't notice the imperfect static factory
method placement.

I'll fix that up shortly. thanks for the think time!

**Comment by adriancole** [comment]
todo add the ability to control the reporter metrics handler

**Comment by adriancole** [comment]
did some interface changes and hit a workload timeout as I spent a lot of energy on email from one person on zipkin-user last night on this project and also can't spend all day today. I will work more later unless someone else wants to carry forward what's begun here. it needs unit tests.

**Comment by adriancole** [comment]
PS I totally forgot we had a note about how to log metrics to the console in the README!

### Programmatic

You can also configure zipkin tracers programmatically. Here's an example:

```java
HttpZipkinTracer.Config config = HttpZipkinTracer.Config.builder()
    // The frontend makes a sampling decision (via Trace.letTracerAndId) and propagates it downstream.
    // This property says sample 100% of traces.
    .initialSampleRate(1.0f)
    // All servers need to point to the same zipkin transport
    .host("127.0.0.1:9411").build();

Tracer tracer = HttpZipkinTracer.create(config,
    // print stats about zipkin to the console
    new JavaLoggerStatsReceiver(Logger.getAnonymousLogger()));
```

**Pull request merged**